src/main/webapp/script.js

```diff
@@ -0,0 +1,371 @@
+ * (it is implemented in canvas.filter.blur).
+ */
+function getBlurredImage(rectsToBlur, image) {
+  // create result canvas and draw not-blurred image on it
+  const resultCanvas = createCanvasForImage(image);
+  const resultCtx = resultCanvas.getContext('2d');
+  resultCtx.drawImage(image, 0, 0);
+
+  // create canvas to put blurred areas there
+  const blurredRectsCanvas = createCanvasForImage(image);
+  const blurredRectsCtx = blurredRectsCanvas.getContext('2d');
+
+  blurredRectsCtx.globalCompositeOperation = 'destination-in';
+
+  for (const _rect of rectsToBlur) {
+    const rect = new Rect(_rect);
+
+    // blur radius depends on the rect size
+    const blurRadius = getBlurRadius(rect);
+
+    // we will get blurred rectangle from this canvas
+    const hiddenBlurredCanvas = getBlurredCanvas(image, blurRadius);
+    const hiddenBlurredCtx = hiddenBlurredCanvas.getContext('2d');
+
+    // get blurred rectangle from blurred canvas
+    const blurredItem = hiddenBlurredCtx.getImageData(
+        rect.leftX, rect.topY, rect.width, rect.height);
+
+    // put that rectangle on canvas for blurred areas
+    blurredRectsCtx.putImageData(blurredItem, rect.leftX, rect.topY);
+  }
+
+  // create canvas to apply smooth blur edges
+  // without it edges of blurred areas will look very rectangular
+  const shadowedRectsCanvas = createCanvasForImage(image);
+  const shadowedRectsCtx = shadowedRectsCanvas.getContext('2d');
+
+  // put blurred areas on another canvas and aplly smooth edges for them
+  shadowedRectsCtx.drawImage(blurredRectsCanvas, 0, 0);
+  shadowedRectsCtx.shadowColor = 'black';
+  shadowedRectsCtx.shadowBlur = 30;
+  shadowedRectsCtx.globalCompositeOperation = 'destination-in';
+
+  shadowedRectsCtx.shadowBlur = 20;
+  shadowedRectsCtx.drawImage(blurredRectsCanvas, 0, 0);
+  shadowedRectsCtx.shadowBlur = 10;
+  shadowedRectsCtx.drawImage(blurredRectsCanvas, 0, 0);
+
+  // put original image and blurred faces on result canvas
+  resultCtx.clearRect(0, 0, image.width, image.height);
+  resultCtx.drawImage(image, 0, 0);
+  resultCtx.drawImage(shadowedRectsCanvas, 0, 0);
+
+  return resultCanvas;
+}
+
+/**
+ * @param {Image} image
+ * @param {Number} blurRadius Parameter which indicates the amount of blur.
+ * @return {HTMLCanvasElement} canvas with blurred image on it.
+ */
+function getBlurredCanvas(image, blurRadius) {
+  const blurredCanvas = createCanvasForImage(image);
+  const blurredCtx = blurredCanvas.getContext('2d');
+
+  blurredCtx.filter = `blur(${blurRadius}px)`;
+  blurredCtx.drawImage(image, 0, 0);
+
+  return blurredCanvas;
+}
+
+/**
+ * Blur radius defines how many pixels on the screen blend into each other.
+ * Here we get the best (to my mind) blur radius for the
+ * dimensions of the rectangle we are going to blur.
+ * @param {Rect} rect
+ * @return {Number} blur radius
+ */
+function getBlurRadius(rect) {
+  if (rect.w > 300) {
+    return 120;
+  }
+  if (rect.w < 30) {
+    return 9;
+  }
+  return 30;
+}
```

Comment:
It is typically easier to read if the conditions are sorted in some way. For example, if you put it this way
```
if (rect.w < 30) {
  return 9;
}
if (rect.w <= 300) {
  return 30;
}
return 120;
```
I'd argue it would be mentally easier to process.